File: packages/maps/tests/test_lint.py
"""Tests for the citation linter (zagent_maps.lint)."""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from zagent_maps.lint import (
    LintReport,
    lint,
    lint_citation,
    load_citations_file,
    main as lint_main,
    parse_citation,
    render_report,
)


@pytest.fixture()
def small_repo(tmp_path):
    (tmp_path / "mod.py").write_text(
        '"""Mod."""\n\nclass C:\n    pass\n\ndef f(a, b):\n    return a + b\n',
        encoding="utf-8",
    )
    (tmp_path / "empty_lines.py").write_text(
        "first\n\nthird\n", encoding="utf-8",
    )
    (tmp_path / "bin.dat").write_bytes(b"\xff\xff\xfe invalid utf8 \xff")
    return tmp_path


# ---------------------------------------------------------------------------
# parse_citation
# ---------------------------------------------------------------------------

def test_parse_single_line():
    c = parse_citation("mod.py:3")
    assert c.path == "mod.py"
    assert c.line_start == 3 and c.line_end == 3


def test_parse_line_range():
    c = parse_citation("mod.py:3-5")
    assert c.line_start == 3 and c.line_end == 5


def test_parse_windows_drive_path():
    c = parse_citation(r"C:\foo\bar.py:12")
    assert c.path == r"C:\foo\bar.py"
    assert c.line_start == 12


def test_parse_rejects_malformed():
    for bad in ["", "mod.py", "mod.py:", ":3", "mod.py:abc", "mod.py:3-", "mod.py:5-3",
               "mod.py:0", "mod.py:-1", "   "]:
        assert parse_citation(bad) is None, f"expected None for {bad!r}"


def test_parse_strips_whitespace():
    c = parse_citation("  mod.py:3  ")
    assert c is not None and c.path == "mod.py" and c.line_start == 3


# ---------------------------------------------------------------------------
# lint_citation
# ---------------------------------------------------------------------------

def test_lint_valid_citation(small_repo):
    c = parse_citation("mod.py:3")
    r = lint_citation(small_repo, c)
    assert r.ok is True
    assert r.error == ""


def test_lint_valid_range(small_repo):
    c = parse_citation("mod.py:3-6")
    r = lint_citation(small_repo, c)
    assert r.ok is True


def test_lint_file_not_found(small_repo):
    c = parse_citation("missing.py:1")
    r = lint_citation(small_repo, c)
    assert r.ok is False and r.error == "file_not_found"


def test_lint_line_out_of_range(small_repo):
    c = parse_citation("mod.py:999")
    r = lint_citation(small_repo, c)
    assert r.ok is False and r.error == "line_out_of_range"
    assert "file has" in r.detail


def test_lint_empty_line(small_repo):
    c = parse_citation("empty_lines.py:2")
    r = lint_citation(small_repo, c)
    assert r.ok is False and r.error == "empty_line"


def test_lint_unreadable_binary_file(small_repo):
    c = parse_citation("bin.dat:1")
    r = lint_citation(small_repo, c)
    assert r.ok is False and r.error == "unreadable_file"


def test_lint_directory_not_a_file(small_repo):
    (small_repo / "dir").mkdir()
    c = parse_citation("dir:1")
    r = lint_citation(small_repo, c)
    assert r.ok is False and r.error == "not_a_file"


# ---------------------------------------------------------------------------
# lint (batch)
# ---------------------------------------------------------------------------

def test_lint_batch_counts(small_repo):
    cits = ["mod.py:3", "mod.py:999", "missing.py:1", "mod.py:abc", "empty_lines.py:2"]
    report = lint(str(small_repo), cits)
    assert isinstance(report, LintReport)
    assert report.total == 5
    assert report.valid == 1
    assert report.invalid == 4
    assert report.passed is False
    assert len(report.results) == 5
    assert report.results[0].ok is True
    assert report.results[1].error == "line_out_of_range"
    assert report.results[2].error == "file_not_found"
    assert report.results[3].error == "malformed_citation"
    assert report.results[4].error == "empty_line"


def test_lint_all_valid_passes(small_repo):
    report = lint(str(small_repo), ["mod.py:3", "mod.py:4", "mod.py:6"])
    assert report.valid == 3 and report.invalid == 0
    assert report.passed is True


def test_lint_as_dict_shape(small_repo):
    report = lint(str(small_repo), ["mod.py:3"])
    d = report.as_dict()
    assert d["total"] == 1 and d["valid"] == 1 and d["invalid"] == 0
    assert d["passed"] is True
    assert d["results"][0]["citation"] == "mod.py:3"
    assert d["results"][0]["ok"] is True


# ---------------------------------------------------------------------------
# load_citations_file
# ---------------------------------------------------------------------------

def test_load_citations_array(tmp_path):
    f = tmp_path / "c.json"
    f.write_text(json.dumps(["a.py:1", "b.py:2-3"]), encoding="utf-8")
    assert load_citations_file(str(f)) == ["a.py:1", "b.py:2-3"]


def test_load_citations_object(tmp_path):
    f = tmp_path / "c.json"
    f.write_text(json.dumps({"citations": ["a.py:1"]}), encoding="utf-8")
    assert load_citations_file(str(f)) == ["a.py:1"]


def test_load_citations_bad_shape(tmp_path):
    f = tmp_path / "c.json"
    f.write_text(json.dumps({"nope": 1}), encoding="utf-8")
    with pytest.raises(ValueError):
        load_citations_file(str(f))


# ---------------------------------------------------------------------------
# render_report
# ---------------------------------------------------------------------------

def test_render_report_lists_invalid(small_repo):
    report = lint(str(small_repo), ["mod.py:3", "missing.py:1"])
    out = render_report(report)
    assert "citation lint report" in out
    assert "passed=False" in out
    assert "OK " in out and "XX " in out


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def test_lint_cli_exit_zero_when_all_valid(small_repo, tmp_path):
    cits = tmp_path / "cits.json"
    cits.write_text(json.dumps(["mod.py:3"]), encoding="utf-8")
    rc = lint_main([str(small_repo), str(cits)])
    assert rc == 0


def test_lint_cli_exit_nonzero_on_invalid(small_repo, tmp_path, capsys):
    cits = tmp_path / "cits.json"
    cits.write_text(json.dumps(["mod.py:999", "missing.py:1"]), encoding="utf-8")
    rc = lint_main([str(small_repo), str(cits)])
    assert rc == 1
    out = capsys.readouterr().out
    assert "invalid=2" in out


def test_lint_cli_json_output(small_repo, tmp_path, capsys):
    cits = tmp_path / "cits.json"
    cits.write_text(json.dumps(["mod.py:3"]), encoding="utf-8")
    rc = lint_main([str(small_repo), str(cits), "--json"])
    assert rc == 0
    data = json.loads(capsys.readouterr().out)
    assert data["valid"] == 1

File: packages/maps-ts/tests/test_lint.mjs
// Unit tests for the TS citation linter (node:test runner).
import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, mkdirSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";

import { parseCitation, lint, lintCitation, renderReport } from "../src/lint.js";

function smallRepo() {
  const tmp = mkdtempSync(join(tmpdir(), "collegium-ts-lint-"));
  writeFileSync(join(tmp, "mod.ts"),
    "/** Mod. */\nexport class C {\n  run(): void {}\n}\nexport function f(a: number, b: number): number {\n  return a + b;\n}\n",
    "utf8");
  writeFileSync(join(tmp, "empty_lines.ts"), "first\n\nthird\n", "utf8");
  // invalid UTF-8 so readFileSync utf8 throws -> unreadable
  writeFileSync(join(tmp, "bin.dat"), Buffer.from([0x00, 0xff, 0xfe, 0x62, 0x69, 0x6e]));
  mkdirSync(join(tmp, "dir"));
  return tmp;
}

test("parseCitation single line", () => {
  const c = parseCitation("mod.ts:3");
  assert.equal(c.path, "mod.ts");
  assert.equal(c.line_start, 3);
  assert.equal(c.line_end, 3);
});

test("parseCitation line range", () => {
  const c = parseCitation("mod.ts:3-5");
  assert.equal(c.line_start, 3);
  assert.equal(c.line_end, 5);
});

test("parseCitation windows drive path", () => {
  const c = parseCitation("C:\\foo\\bar.ts:12");
  assert.equal(c.path, "C:\\foo\\bar.ts");
  assert.equal(c.line_start, 12);
});

test("parseCitation rejects malformed", () => {
  for (const bad of ["", "mod.ts", "mod.ts:", ":3", "mod.ts:abc", "mod.ts:3-", "mod.ts:5-3", "mod.ts:0", "   "]) {
    assert.equal(parseCitation(bad), null, `expected null for ${JSON.stringify(bad)}`);
  }
});

test("parseCitation strips whitespace", () => {
  const c = parseCitation("  mod.ts:3  ");
  assert.equal(c.path, "mod.ts");
  assert.equal(c.line_start, 3);
});

test("lintCitation valid", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("mod.ts:3"));
  assert.equal(r.ok, true);
  assert.equal(r.error, "");
});

test("lintCitation valid range", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("mod.ts:3-6"));
  assert.equal(r.ok, true);
});

test("lintCitation file not found", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("missing.ts:1"));
  assert.equal(r.ok, false);
  assert.equal(r.error, "file_not_found");
});

test("lintCitation line out of range", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("mod.ts:999"));
  assert.equal(r.ok, false);
  assert.equal(r.error, "line_out_of_range");
  assert.ok(r.detail.includes("file has"));
});

test("lintCitation empty line", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("empty_lines.ts:2"));
  assert.equal(r.ok, false);
  assert.equal(r.error, "empty_line");
});

test("lintCitation unreadable binary file", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("bin.dat:1"));
  assert.equal(r.ok, false);
  assert.equal(r.error, "unreadable_file");
});

test("lintCitation directory not a file", () => {
  const repo = smallRepo();
  const r = lintCitation(repo, parseCitation("dir:1"));
  assert.equal(r.ok, false);
  assert.equal(r.error, "not_a_file");
});

test("lint batch counts", () => {
  const repo = smallRepo();
  const report = lint(repo, ["mod.ts:3", "mod.ts:999", "missing.ts:1", "mod.ts:abc", "empty_lines.ts:2"]);
  assert.equal(report.total, 5);
  assert.equal(report.valid, 1);
  assert.equal(report.invalid, 4);
  assert.equal(report.passed, false);
  assert.equal(report.results[0].ok, true);
  assert.equal(report.results[1].error, "line_out_of_range");
  assert.equal(report.results[2].error, "file_not_found");
  assert.equal(report.results[3].error, "malformed_citation");
  assert.equal(report.results[4].error, "empty_line");
});

test("lint all valid passes", () => {
  const repo = smallRepo();
  const report = lint(repo, ["mod.ts:3", "mod.ts:4", "mod.ts:6"]);
  assert.equal(report.valid, 3);
  assert.equal(report.invalid, 0);
  assert.equal(report.passed, true);
});

test("renderReport lists invalid", () => {
  const repo = smallRepo();
  const out = renderReport(lint(repo, ["mod.ts:3", "missing.ts:1"]));
  assert.ok(out.includes("citation lint report"));
  assert.ok(out.includes("passed=false"));
  assert.ok(out.includes("OK "));
  assert.ok(out.includes("XX "));
});

test("cli lint exits non-zero on invalid", () => {
  const repo = smallRepo();
  const citsFile = join(repo, "cits.json");
  writeFileSync(citsFile, JSON.stringify(["mod.ts:999", "missing.ts:1"]), "utf8");
  const cli = join(dirname(fileURLToPath(import.meta.url)), "..", "src", "cli.js");
  const r = spawnSync(process.execPath, [cli, "lint", repo, citsFile], { encoding: "utf8" });
  assert.equal(r.status, 1);
  assert.ok(r.stdout.includes("invalid=2"));
});
